File: src/SK-RM/art_trace_info.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

// 定義 tracer 結構
typedef struct {
    FILE *trace_file;
    int enable;
    int domain_shift;
    int track_shift;
    int trace_count;
} tracer;

// 初始化
void tracer_init(tracer *t, const char *filename) {
    t->enable = 1;
    t->domain_shift = 6;
    t->track_shift = t->domain_shift + (int)(log2(4096)); // default 12
    t->trace_count = 0;

    t->trace_file = fopen(filename, "a");
    if (t->trace_file) {
        fprintf(t->trace_file, "NVMV 2\n");
        printf("[trace enabled]\n");
    }
}

// 設定 domain
void tracer_set_domain(tracer *t, int domain) {
    t->track_shift = t->domain_shift + (int)(log2(domain));
}

// 關閉 trace
void tracer_close(tracer *t) {
    if (t->trace_file) {
        fclose(t->trace_file);
    }
}

// address 計算
uint64_t tracer_calc_address(tracer *t, uint32_t track, uint32_t domain) {
    return ((uint64_t)track << t->track_shift) + ((uint64_t)domain << t->domain_shift);
}

// 字串轉 hex，填充 128 個字元
void str_to_hex(const char *value, char *hex_out) {
    if (!value || strlen(value) == 0) {
        memset(hex_out, '0', 128);
        hex_out[128] = '\0';
        return;
    }

    char bin_str[1024] = {0};
    for (size_t i = 0; i < strlen(value); i++) {
        char bin[9];
        snprintf(bin, sizeof(bin), "%08b", (unsigned char)value[i]);
        strcat(bin_str, bin);
    }

    // 將 binary string 轉 hex string
    unsigned char tmp[256] = {0};
    size_t bin_len = strlen(bin_str);
    for (size_t i = 0; i < bin_len; i += 4) {
        char chunk[5] = {0};
        strncpy(chunk, bin_str + i, 4);
        int val = (int)strtol(chunk, NULL, 2);
        sprintf((char *)tmp + strlen((char *)tmp), "%x", val);
    }

    // 填充長度
    strncpy(hex_out, (char *)tmp, 128);
    size_t tmp_len = strlen(hex_out);
    if (tmp_len < 128) {
        memset(hex_out + tmp_len, '0', 128 - tmp_len);
    }
    hex_out[128] = '\0';
}

// skyrmion insert
void tracer_skyrmion_insert(tracer *t, uint32_t track, uint32_t domain, const char *newvalue, int cycle) {
    if (!t->enable || !t->trace_file) return;

    uint64_t addr = tracer_calc_address(t, track, domain);

    char new_hex[129];
    str_to_hex(newvalue, new_hex);
    char zero[129];
    memset(zero, '0', 128);
    zero[128] = '\0';

    fprintf(t->trace_file, "%d I 0x%lX %s %s 0\n", cycle, addr, new_hex, zero);
}

// skyrmion delete
void tracer_skyrmion_delete(tracer *t, uint32_t track, uint32_t domain, const char *oldvalue, int cycle) {
    if (!t->enable || !t->trace_file) return;

    uint64_t addr = tracer_calc_address(t, track, domain);

    char old_hex[129];
    str_to_hex(oldvalue, old_hex);
    char zero[129];
    memset(zero, '0', 128);
    zero[128] = '\0';

    fprintf(t->trace_file, "%d D 0x%lX %s %s 0\n", cycle, addr, zero, old_hex);
}

// skyrmion read
void tracer_skyrmion_read(tracer *t, uint32_t track, uint32_t domain, const char *value, int cycle) {
    if (!t->enable || !t->trace_file) return;

    uint64_t addr = tracer_calc_address(t, track, domain);

    char hex_val[129];
    str_to_hex(value, hex_val);

    fprintf(t->trace_file, "%d R 0x%lX %s %s 0\n", cycle, addr, hex_val, hex_val);
}

// skyrmion write
void tracer_skyrmion_write(tracer *t, uint32_t track, uint32_t domain, const char *oldvalue, const char *newvalue, int cycle) {
    if (!t->enable || !t->trace_file) return;

    uint64_t addr = tracer_calc_address(t, track, domain);

    char old_hex[129];
    str_to_hex(oldvalue, old_hex);
    char new_hex[129];
    str_to_hex(newvalue, new_hex);

    fprintf(t->trace_file, "%d W 0x%lX %s %s 0\n", cycle, addr, new_hex, old_hex);
}

// 測試用
int main() {
    tracer t;
    tracer_init(&t, "trace_output.nvt");

    tracer_skyrmion_read(&t, 1, 64, "read_value", 100);
    tracer_skyrmion_write(&t, 1, 1024, "old_value", "new_value", 101);
    tracer_skyrmion_insert(&t, 1, 1024, "data_value", 102);
    tracer_skyrmion_delete(&t, 1, 1024, "data_value", 103);

    tracer_close(&t);

    return 0;
}

File: src/SK-RM/art_trace_info.py
import math


class tracer:
    def __init__(self):
        self.filename = None
        self.trace_file = None
        self._enable = 0
        self.domain_shift = 6  # Default offset, total addr = R:C:RK:BK:CH, RK + BK + CH = 6 bits
        # self.track_shift = self.domain_shift + int(math.log2(4096))  # DOMAIN_NR from C++, log2(4096)= 12 = C = domain(bit in track)
        self.track_shift = self.domain_shift + int(math.log2(128*8)) # a track has 128B domain = 128*8 bits = 1024 bits
        self.trace_count = 0 
        self.DBC_track = 64   # R = 6 => 2^6 = total 64 tracks on one DBC, track has 6 bits for counting

    def initialize_trace(self, filename):
        self._enable = 1
        self.filename = filename
        self.trace_file = open(self.filename, 'a')
        if self.trace_file:
            self.trace_file.write("NVMV 2\n")
            print("[trace enabled]")

    def set_domain(self, domain):
        self.track_shift = self.domain_shift + int(math.log2(domain))

    def close_trace(self):
        if self.trace_file:
            self.trace_file.close()

    def skyrmion_read(self, track, domain, value, cycle):
        if self._enable:
            address = self.calc_address(track, domain)
            if self.trace_file:
                self.trace_file.write(f"{cycle} R 0x{address:X} {self.str_to_hex(value)} {self.str_to_hex(value)} 0\n")

    def skyrmion_write(self, track, domain, oldvalue, newvalue, cycle):
        if self._enable:
            address = self.calc_address(track, domain)
            if self.trace_file:
                self.trace_file.write(f"{cycle} W 0x{address:X} {self.str_to_hex(newvalue)} {self.str_to_hex(oldvalue)} 0\n")

    def skyrmion_insert(self, track, domain, newvalue, cycle):
        if self._enable:
            address = self.calc_address(track, domain)
            if self.trace_file:
                zero="0"*128
                self.trace_file.write(f"{cycle} I 0x{address:X} {self.str_to_hex(newvalue)} {zero} 0\n")

    def skyrmion_delete(self, track, domain, oldvalue, cycle):
        if self._enable:
            address = self.calc_address(track, domain)
            if self.trace_file:
                zero="0"*128
                self.trace_file.write(f"{cycle} D 0x{address:X} {zero} {self.str_to_hex(oldvalue)} 0\n")

    def calc_address(self, track, domain):
        address = (track << self.track_shift) + (domain << self.domain_shift)
        # open("current_domain.txt", mode='a').write(f"{track} {domain}\n")

        return address

    def str_to_hex(self, value):
        if self._enable:
            if isinstance(value, int):
                hex_value = format(value, 'x')
                return hex_value.ljust(128, '0')

            if value == "":
                return "0" * 128

            binary_parts = [bin(ord(char))[2:].zfill(8) for char in value]
            binary_string = ''.join(binary_parts)
            hex_string = hex(int(binary_string, 2))[2:]
            return hex_string.ljust(128, '0')

if __name__ == "__main__":
    skyrmion_mem = tracer()
    skyrmion_mem.initialize_trace("trace_output.nvt")
    skyrmion_mem.skyrmion_read(1, 64, "read_value", 100)  # Simulated read operation
    skyrmion_mem.skyrmion_write(1, 1024, "old_value", "new_value", 101)  # Simulated write operation
    skyrmion_mem.skyrmion_insert(1, 1024, "data_value", 102)  # Simulated insert operation
    skyrmion_mem.skyrmion_delete(1, 1024, "data_value", 103)  # Simulated delete operation
    skyrmion_mem.close_trace()
